src/include/daos_srv/dtx_srv.h

```diff
@@ -64,7 +67,13 @@ struct dtx_handle {
 					 /* Modified shared items: object/key */
 					 dth_modify_shared:1,
 					 /* The DTX entry is in active table. */
-					 dth_active:1;
+					 dth_active:1,
+					 /* Leader oid is touched. */
+					 dth_touched_leader_oid:1,
+					 /* Local TX is started. */
+					 dth_local_tx_started:1,
+					 /* The last sub-modification for DTX */
+					 dth_last_modification:1;
```

Comment:
I'm wondering how will do the error cleanup on transaction abort? For example, there are 3 vos operations, first one started transaction, and third operation failed and aborted transaction, how will we do the error cleanup for first two operations? (assume there are some state needs be cleared when the whole transaction aborted, like free memory, revert read timestamp, cancel reservation, etc.)

Reply:
There are three kinds of things to be recovered when the DTX failed and aborted:

1) The on-disk modifications that will be rollback by PMDK automatically.

2) DTX index in DRAM. That will be cleanup by vos_dtx_cleanup_dth().

3) VOS object cache. That will be evicted during vos_update_end()/vos_obj_punch() if the modification is failed.

As for the read timestamp, I am not sure, but I think it may be OK if we just keep them there. Since the TS is increased only, the worst case is that it may cause some fake DTX conflict and restart, but no correctness issue.

Reply:
Just realize that our current evict object on DTX abort only handle current object. But if there are multiple objects created by the DTX, we need to evict all of them. I will refresh the patch for that.